pkg/env/certificateRefresher.go

```diff
@@ -0,0 +1,111 @@
+package env
+
+import (
+	"context"
+	"crypto/rsa"
+	"crypto/x509"
+	"sync/atomic"
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
+
+type CertGetter func(context.Context, string) (*rsa.PrivateKey, []*x509.Certificate, error)
+
+type RefreshingCertificateManager interface {
+	StartPoller(context.Context) error
+	StopPoller()
+	GetCertificates() (*rsa.PrivateKey, []*x509.Certificate)
+}
+
+type refreshingCertificate struct {
+	certs    atomic.Value
+	key      atomic.Value
+	interval time.Duration
+	logger   *logrus.Entry
+	getter   func(context.Context) (*rsa.PrivateKey, []*x509.Certificate, error)
+	done     chan bool
+}
+
+func NewRefreshingCertificate(ctx context.Context, logger *logrus.Entry, interval time.Duration, getter CertGetter, certificateName string) *refreshingCertificate {
+	r := &refreshingCertificate{
+		logger:   logger,
+		interval: interval,
+		getter: func(ctx context.Context) (*rsa.PrivateKey, []*x509.Certificate, error) {
+			return getter(ctx, certificateName)
+		},
+		done: make(chan bool),
+	}
+
+	return r
+}
+
+func (r *refreshingCertificate) StartPoller(ctx context.Context) error {
+	// initial pull to get the certificate start
+	err := r.pullCertificate(ctx)
+	if err != nil {
+		return err
+	}
+
+	r.pollCertificate(ctx, r.done)
+
+	return nil
+}
+
+func (r *refreshingCertificate) StopPoller() {
+	close(r.done)
+}
```

Comment:
Is this used outside the tests?

Reply:
It was part of the concept. I have exported the stop channel out to allow for proper termination with the RP.